extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+    } else if (preconnect) {
+      preconnect.forEach(p => {
+        this.preconnect.url(p, true);
+      });
+    }
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    if (this.intersectionObserver_) {
+      this.intersectionObserver_.fire();
+    }
+    if (this.layoutMeasureExecuted_ || !verifySignatureIsAvailable()) {
+      // onLayoutMeasure gets called multiple times.
+      return;
+    }
+    this.layoutMeasureExecuted_ = true;
+    if (isPositionFixed(this.element, this.getWin())) {
+      user.error('<amp-ad> is not allowed to be placed in elements with ' +
+                 'position:fixed: %s', this.element);
+      return;
+    }
+    // OnLayoutMeasure can be called when page is in prerender so delay until
+    // visible.  Assume that it is ok to call isValidElement as it should
+    // only being looking at window, immutable properties (i.e. location) and
+    // its element ancestry.
+    if (!this.isValidElement()) {
+      // TODO(kjwright): collapse?
+      user.warn('Amp Ad', 'Amp ad element ignored as invalid', this.element);
+      return;
+    }
+
+    // Increment unique promise ID so that if its value changes within the
+    // promise chain due to cancel from unlayout, the promise will be rejected.
+    this.promiseId_++;
+    const promiseId = this.promiseId_;
+    this.adPromise_ = viewerFor(this.getWin()).whenFirstVisible()
+      .then(() => {
+        if (promiseId != this.promiseId_) {
+          return Promise.reject(cancellation());
+        }
+        return this.getAdUrl();
+      })
+      .then(adUrl => {
+        if (promiseId != this.promiseId_) {
+          return Promise.reject(cancellation());
+        }
+        this.adUrl_ = adUrl;
+        return this.sendXhrRequest_(adUrl);
+      })
+      .then(fetchResponse => {
+        if (promiseId != this.promiseId_) {
+          return Promise.reject(cancellation());
+        }
+        if (fetchResponse && fetchResponse.arrayBuffer) {
+          return fetchResponse.arrayBuffer().then(
+            bytes => {
+              if (promiseId != this.promiseId_) {
+                return Promise.reject(cancellation());
+              }
+              return this.validateAdResponse_(fetchResponse, bytes)
+                .then(valid => {
+                  if (promiseId != this.promiseId_) {
+                    return Promise.reject(cancellation());
+                  }
+                  return this.maybeRenderAmpAd_(valid, bytes);
+                });
+            });
+        } else {
+          return Promise.resolve(false);
+        }
+      })
+      .catch(error => this.promiseErrorHandler_(error));
+  }
+
+  /**
+   * Handles uncaught errors within promise flow.
+   * @param {string|Error} error
+   * @return {!Promise<string>}
+   * @private
+   */
+  promiseErrorHandler_(error) {
+    if (error && error.message && error.message.indexOf('amp-a4a: ') == 0) {
+      // caught previous call to promiseErrorHandler?  Infinite loop?
```

Comment:
This looks to be because `#adPromise` already has a `catch(this.promiseErrorHandler_)` block, it rejects, and you then add another `catch(this.promiseErrorHandler_)` block inside `#layoutCallback`.